Move dev-related files and instructions to dev/ directory; add committer's instructions

leventov opened the pull request:
Adds `committer-instructions.md` file for Druid committers. For now, it has a PR action item checklist, i. e. a list of actions that a committer should do with their own PRs or a PR from a non-committer that they encounter.

Open question: what about PRs from new contributors? Should they sign something like Apache ICLA, and therefore a committer should suggest the new contributor do this?

In the future, `committer-instructions.md` may be updated with more general "Code of Conduct"  for committers, an updated version of "General committer guidelines" from this page: http://druid.io/community/

jon-wei commented:
> Open question: what about PRs from new contributors? Should they sign something like Apache ICLA, and therefore a committer should suggest the new contributor to do this?

For ICLA, only the committers need that: https://lists.apache.org/thread.html/32c3f6aaa646da8585d3e8ede1fa87468e82fa064731463232081355@%3Cdev.druid.apache.org%3E

gianm commented on committer_readme.md:
```diff
@@ -0,0 +1,56 @@
+     - **`Design Review`** - for changes that will be hard to undo after they appear in some Druid release, and/or
+     changes that will have lasting consequences in the codebase. Examples:
+        - Major architectural changes or API changes
+        - Adding new or changing behaviour of existing query types (e. g. changing and algorithm behind some query type
```
Should 'and' be 'an'?

leventov replied:
Thanks, fixed

gianm commented on committer_readme.md:
```diff
@@ -0,0 +1,56 @@
+     - Consider adding one or several **`Area -`** tags. Consider creating a new `Area -` tag if none of the existing
+     `Area` tags is applicable to the PR.
+
+2. Consider adding `Bug` and `Security` PRs to the next Druid milestone. Don't add PRs that are neither `Bug` nor
```
Add some rationale? Maybe:

> Consider adding any `Bug` and `Security` PRs to the next Druid milestone whenever they are important enough to fix before the next release. This ensures that they will be considered by the next release manager as potential release blockers. Please don't add PRs that are neither `Bug` nor `Security`-related to milestones until after they are committed, to avoid cluttering the release manager's workflow.

leventov replied:
Thanks, updated

gianm commented on committer_readme.md:
```diff
@@ -0,0 +1,56 @@
+     configuration options, emitted metric names, HTTP endpoint paths and parameters that are added or changed in the
+     PR. If they are not listed, ask the PR author to update the PR description.
+
+     - **`Incompatible`** - for changes that may change the behaviour of Druid clusters if a new version of Druid with
```
IMO we should loosen a bit from "may change the behaviour of Druid clusters". Lots of changes may change the behavior of Druid clusters in various ways (minor perf changes, memory usage changes, & so on) but not all of them are really incompatible. How about a definition like:

> Changes that alter public API elements (`@PublicApi` or `@ExtensionPoint`), runtime configuration options, emitted metric names, HTTP endpoint behavior, or server behavior in some way that affects one of the following:
>
> - Ability to do a rolling update as documented on http://druid.io/docs/latest/operations/rolling-updates.html without needing any modifications to server configurations or query workload.
> - Ability to roll back a Druid cluster to a prior version.
> - Ability to continue using old Druid extensions without recompiling them.
>
> If a Druid release contains any "Incompatible" changes then it should have a second-digit version bump (i.e. from 0.10 -> 0.11).
>
> Note that no matter what, we must support the ability to do a rolling update _somehow_ (even if some special care is needed), and the ability to roll back to at least the immediate prior Druid version. If a change makes either one of these impossible then it should be re-designed.

leventov replied:
Applied. Moved the part about the version bump to a new "PR merge action item checklist".

Left the last note, although it seems to me that it really belongs to a code review checklist rather than a PR action item checklist, because it may require some thoughtful analysis of the PR. Also, it should be considered by the PR author (even non-committer) themselves during their PR self-review.

asdf2014 commented on committer_readme.md:
```diff
@@ -0,0 +1,94 @@
+        - Adding new or changing existing emitted metrics
+        - Other major changes
+
+     The PR description should succintly, but completely list all public API elements (@PublicApi or @ExtensionPoint),
```
succintly -> succinctly

leventov replied:
Thanks, fixed

leventov commented:
@gianm @asdf2014 I've extended the scope of this PR. I've moved some files from root to a new `dev/` directory. I've also added more committer's instructions:
 - PR merge action item checklist
 - Creating a new tag on Github

leventov commented:
@gianm @asdf2014 do you have more comments?

yurmix commented:
@leventov looks nice! two small suggestions:
1. There is no hyperlink to `committer-instructions.md`, I think `CONTRIBUTING.md` should link to it.
2. I think the asset files should be organized into sub-folders (perhaps `TeamCity_images`, `IDE_images`), especially since there is no README.md in the folder so it's harder to sort things out.

leventov commented:
@yurmix good suggestions, thanks. I've moved images to a folder. I didn't mention `committer-instructions.md` in `CONTRIBUTING.md` because committers don't read CONTRIBUTING.md, but I referenced the `dev/` directory in general. I think it should be the place for all development-related prose, like checklists, code style description, [principles for choosing logging severity](https://github.com/apache/incubator-druid/issues/7362), etc.

leventov commented:
@gianm do you have more comments?

leventov commented:
@gianm do you want to leave more comments?
